### PR TITLE
process/util use f.Readdirnames to list pids

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

Speed up listing process by
 * using f.Readdirnames(), this avoid doing lstat() per entry
 * not sorting the directory by names implemented in Readdir()

### Motivation

quick benchmark

```
BenchmarkOldWithAllProcs
BenchmarkOldWithAllProcs-16    	    3429	    328718 ns/op	   87470 B/op	    1233 allocs/op
BenchmarkWithAllProcs
BenchmarkWithAllProcs-16       	    5763	    215248 ns/op	   56332 B/op	     682 allocs/op
BenchmarkAllPidsProcs
BenchmarkAllPidsProcs-16       	    5656	    208567 ns/op	   41296 B/op	     671 allocs/op


```
### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/process/util/util.go [hunk 1]
```diff
@@ -127,29 +127,42 @@ func GetSysRoot() string {
...
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	dirs, err := f.Readdirnames(-1)
+	if err != nil {
+		return nil, err
+	}
+
+	pids := make([]int, len(dirs))
```
Comment (guyarb) [inline pkg/process/util/util.go:142]:
```suggestion
	pids := make([]int, 0, len(dirs))
```

File: pkg/process/util/util.go [hunk 2]
```diff
@@ -127,29 +127,42 @@ func GetSysRoot() string {
...
+	for _, name := range dirs {
+		var pid int
+		if pid, err = strconv.Atoi(name); err != nil {
+			continue
+		}
```
Comment (guyarb) [inline pkg/process/util/util.go:147]:
```suggestion
		if pid, err := strconv.Atoi(name); err != nil {
			continue
		}
```

**Review by guyarb — CHANGES_REQUESTED** [review]
small fixes
please add memory benchmarks to the pr description

**Review by guyarb — CHANGES_REQUESTED** [review]
small fixes
please add memory benchmarks to the pr description

**Review by guyarb — CHANGES_REQUESTED** [review]
small fixes
please add memory benchmarks to the pr description

**Review by guyarb — APPROVED** [review]

File: pkg/process/util/util.go [hunk 3]
```diff
@@ -127,29 +127,41 @@ func GetSysRoot() string {
...
+	}
+	defer f.Close()
+	dirs, err := f.Readdirnames(-1)
+	if err != nil {
+		return nil, err
+	}
+
+	pids := make([]int, 0, len(dirs))
+	for _, name := range dirs {
+		if pid, err := strconv.Atoi(name); err != nil {
```
Comment (guyarb) [inline pkg/process/util/util.go:144]:
```suggestion
		if _, err := strconv.Atoi(name); err != nil {
```

File: pkg/process/util/util.go [hunk 4]
```diff
@@ -127,29 +127,41 @@ func GetSysRoot() string {
...
+			continue
+		}
+		pids = append(pids, pid)
```
Comment (guyarb) [inline pkg/process/util/util.go:147]:
```suggestion
		if pid, err := strconv.Atoi(name); err == nil {
			pids = append(pids, pid)
		}
		
```

Comment (guyarb) [reply pkg/process/util/util.go:147]: otherwise there are compilation error

**Review by guyarb — APPROVED** [review]

File: pkg/process/util/util.go [hunk 5]
```diff
@@ -127,29 +127,40 @@ func GetSysRoot() string {
...
+	}
+	defer f.Close()
+	dirs, err := f.Readdirnames(-1)
+	if err != nil {
+		return nil, err
+	}
+
+	pids := make([]int, 0, len(dirs))
+	for _, name := range dirs {
+		if pid, err := strconv.Atoi(name); err == nil {
```
Comment (mbotarro) [inline pkg/process/util/util.go:144]: Does `Readdirnames` still return `.` and `..`?

I wonder if we should keep this test here:
```golang
if !f.IsDir() || f.Name() == "." || f.Name() == ".." {
	continue
}
```

**Review by mbotarro — APPROVED** [review]

File: pkg/process/util/util.go (hunk 5 above)
Comment (nplanel) [reply pkg/process/util/util.go:144]: Readdirnames doesn't return `.` and `..` entries https://cs.opensource.google/go/go/+/master:src/os/dir_unix.go;l=107?q=readdir&ss=go%2Fgo:src%2Fos%2F

Comment (mbotarro) [reply pkg/process/util/util.go:144]: Thanks for clarifying it!

**Pull request merged**